torch/utils/data/dataloader.py

```diff
@@ -1489,17 +1494,23 @@ def _try_put_index(self):
         for _ in range(self._num_workers):  # find the next active worker, if any
             worker_queue_idx = next(self._worker_queue_idx_cycle)
             if self._workers_status[worker_queue_idx]:
-                break
+                if self._in_order:
+                    break
+                elif self._workers_num_tasks[worker_queue_idx] < self._prefetch_factor:
```

Comment:
For the case of IterableDatasets where some workers are finished, will this change prefetching behaviour? Suppose we have 2 workers, prefetch_factor = 2, and worker0 has raised stop iteration. With `in_order == True` how what's the maximum number of tasks worker1 will have at a time, is it 2 or 4? With this change, and in_order == False, worker1 will be limited to 2 outstanding tasks

Reply:
Good question! From a quick check it will.
With in_order == True, worker1 can have up to 4 tasks in it's queue in that situation, whereas with in_order == False it will be limited to 2.
Instead of the check being `< self._prefetch_factor` it could be changed to account for the number of _active_ remaining workers instead? e.g. something like
```python
< math.floor(self._prefetch_factor*(self._num_workers/sum(self._workers_status))
```
(It being 2 does align slightly more with what my initial understanding of prefecth_factor was, but with IterableDatasets is a bit different)
Let me know what you think!

Reply:
that LGTM

Reply:
Sorry, I'm not sure I follow why your suggestion would work.
From my understanding, starting with say 5 workers and prefetch = 2, that's a maximum of 10 tasks at a time.
If it then goes down to 3 active workers, my suggested change above would give 5*2 // 3 = 3 as the new prefetch factor for each worker. That would distribute work as uniformly as possible but still keep distributing if one or two workers were slower. (`sum([True])` also gives 1 so I'm not sure the 1-status is required?)

Reply:
doh, total brain fart, that's what I get for commenting before I've had my coffee, not exactly sure what I was thinking. It seems like in the worst case, we may prefetch fewer tasks by up to num_workers, but we only get close to that bound when num_workers is large. This seems reasonable to me

Reply:
No worries! Updated now 👍